<div style="color:#006666; padding:0px 10px; border-radius:5px; font-size:18px;text-align:center""><h1 style='margin:10px 5px'>Pandas Project</h1>
</div>

<hr>
<div style="color:#006666; text-align:right">
© Copyright Machine Learning Plus     
<br>    
Visit machinelearningplus.com for more information.
</div>

<div class="alert alert-info" style="background-color:#006666; color:white; padding:0px 10px; border-radius:5px;"><h2 style='margin:10px 5px'>1. Data</h2>
</div>

The data is hosted by the Global Health Observatory (GHO) data repository under World Health Organization (WHO) which keeps track of the health status as well as many other related factors for all countries.

The dataset related to life expectancy, health factors for 193 countries has been collected from the same WHO data repository website and its corresponding economic data was collected from United Nation website.

Each row in the dataset represents a country-year pair and corresponding health and economic factors. 

|Field|Description|
|---:|:---|
|Country|Country|
|Year|Year|
|Status|Developed or Developing status|
|Life expectancy|Life Expectancy in age|
|Adult Mortality|Adult Mortality Rates of both sexes (probability of dying between 15 and 60 years per 1000 population)|
|infant deaths|Number of Infant Deaths per 1000 population|
|Alcohol|Alcohol, recorded per capita (15+) consumption (in litres of pure alcohol)|
|percentage expenditure|Expenditure on health as a percene of Gross Domestic Product per capita(%)|
|Hepatitis B|Hepatitis B (HepB) immunization coverage among 1-year-olds (%)|
|Measles|Measles - number of reported cases per 1000 population|
|BMI|Average Body Mass Index of entire population|
|under-five deaths|Number of under-five deaths per 1000 population|
|Polio|Polio (Pol3) immunization coverage among 1-year-olds (%)|
|Total expenditure|General government expenditure on health as a percene of total government expenditure (%)|
|Diphtheria|Diphtheria tetanus toxoid and pertussis (DTP3) immunization coverage among 1-year-olds (%)|
|HIV/AIDS|Deaths per 1 000 live births HIV/AIDS (0-4 years)|
|GDP|Gross Domestic Product per capita (in USD)|
|Population|Population of the country|
|thinness 1-19 years|Prevalence of thinness among children and adolescents for Age 10 to 19 (%)|
|thinness 5-9 years|Prevalence of thinness among children for Age 5 to 9(%)|
|Income composition of resources|Income composition of resources|
|Schooling|Number of years of Schooling(years)|

<div class="alert alert-info" style="background-color:#006666; color:white; padding:0px 10px; border-radius:5px;"><h2 style='margin:10px 5px'>2. Tasks</h2>
</div>

1. Display the first five rows and make it show all the columns without truncating.
2. Display all the columns and their types
3. Extract the number of rows and columns present in the data
4. Does any of columns have extra spaces in the name. If so, remove the extra spaces
5. Get the summary of the data
6. Are there any missing values in the data? If so, which column has the 2nd most missing values
7. Treat the missing values
8. Are there any outliers in the 'Life Expectancy' column in the data.
9. How many countries have the status as developed and developing respectively
10. Which country has the highest BMI (average)
11. Who has higher life expectancy, developed or developing?
12. Which year experienced the lowest adult mortality
13. What is the average life expectancy when the GDP is above 600

In [ ]:
# Import libraries
import pandas as pd

In [ ]:
# Read the life expectancy data
df = pd.read_csv('Datasets/Life Expectancy Data.csv')
df

Country  Year      Status  Life expectancy   Adult Mortality  \
0     Afghanistan  2015  Developing              65.0            263.0   
1     Afghanistan  2014  Developing              59.9            271.0   
2     Afghanistan  2013  Developing              59.9            268.0   
3     Afghanistan  2012  Developing              59.5            272.0   
4     Afghanistan  2011  Developing              59.2            275.0   
...           ...   ...         ...               ...              ...   
2933     Zimbabwe  2004  Developing              44.3            723.0   
2934     Zimbabwe  2003  Developing              44.5            715.0   
2935     Zimbabwe  2002  Developing              44.8             73.0   
2936     Zimbabwe  2001  Developing              45.3            686.0   
2937     Zimbabwe  2000  Developing              46.0            665.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   \
0                62     0.01               71.279624         65.0      1154   
1                64     0.01               73.523582         62.0       492   
2                66     0.01               73.219243         64.0       430   
3                69     0.01               78.184215         67.0      2787   
4                71     0.01                7.097109         68.0      3013   
...             ...      ...                     ...          ...       ...   
2933             27     4.36                0.000000         68.0        31   
2934             26     4.06                0.000000          7.0       998   
2935             25     4.43                0.000000         73.0       304   
2936             25     1.72                0.000000         76.0       529   
2937             24     1.68                0.000000         79.0      1483   

      ...  Polio  Total expenditure  Diphtheria    HIV/AIDS         GDP  \
0     ...    6.0               8.16         65.0        0.1  584.259210   
1     ...   58.0               8.18         62.0        0.1  612.696514   
2     ...   62.0               8.13         64.0        0.1  631.744976   
3     ...   67.0               8.52         67.0        0.1  669.959000   
4     ...   68.0               7.87         68.0        0.1   63.537231   
...   ...    ...                ...          ...        ...         ...   
2933  ...   67.0               7.13         65.0       33.6  454.366654   
2934  ...    7.0               6.52         68.0       36.7  453.351155   
2935  ...   73.0               6.53         71.0       39.8   57.348340   
2936  ...   76.0               6.16         75.0       42.1  548.587312   
2937  ...   78.0               7.10         78.0       43.5  547.358878   

      Population   thinness  1-19 years   thinness 5-9 years  \
0     33736494.0                   17.2                 17.3   
1       327582.0                   17.5                 17.5   
2     31731688.0                   17.7                 17.7   
3      3696958.0                   17.9                 18.0   
4      2978599.0                   18.2                 18.2   
...          ...                    ...                  ...   
2933  12777511.0                    9.4                  9.4   
2934  12633897.0                    9.8                  9.9   
2935    125525.0                    1.2                  1.3   
2936  12366165.0                    1.6                  1.7   
2937  12222251.0                   11.0                 11.2   

      Income composition of resources  Schooling  
0                               0.479       10.1  
1                               0.476       10.0  
2                               0.470        9.9  
3                               0.463        9.8  
4                               0.454        9.5  
...                               ...        ...  
2933                            0.407        9.2  
2934                            0.418        9.5  
2935                            0.427      

### A1. Display the first five rows of 'Life Expectancy Data' and make it show all the columns without truncating.

In [ ]:
pd.set_option('display.max_columns', 50)
df.head(5)

Country  Year      Status  Life expectancy   Adult Mortality  \
0  Afghanistan  2015  Developing              65.0            263.0   
1  Afghanistan  2014  Developing              59.9            271.0   
2  Afghanistan  2013  Developing              59.9            268.0   
3  Afghanistan  2012  Developing              59.5            272.0   
4  Afghanistan  2011  Developing              59.2            275.0   

   infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   \
0             62     0.01               71.279624         65.0      1154   
1             64     0.01               73.523582         62.0       492   
2             66     0.01               73.219243         64.0       430   
3             69     0.01               78.184215         67.0      2787   
4             71     0.01                7.097109         68.0      3013   

    BMI   under-five deaths   Polio  Total expenditure  Diphtheria   \
0   19.1                  83    6.0               8.16         65.0   
1   18.6                  86   58.0               8.18         62.0   
2   18.1                  89   62.0               8.13         64.0   
3   17.6                  93   67.0               8.52         67.0   
4   17.2                  97   68.0               7.87         68.0   

    HIV/AIDS         GDP  Population   thinness  1-19 years  \
0        0.1  584.259210  33736494.0                   17.2   
1        0.1  612.696514    327582.0                   17.5   
2        0.1  631.744976  31731688.0                   17.7   
3        0.1  669.959000   3696958.0                   17.9   
4        0.1   63.537231   2978599.0                   18.2   

    thinness 5-9 years  Income composition of resources  Schooling  
0                 17.3                            0.479       10.1  
1                 17.5                            0.476       10.0  
2                 17.7                            0.470        9.9  
3                 18.0                            0.463        9.8  
4                 18.2                            0.454        9.5

### A2. Display all the columns and their types

In [ ]:
# Set options to display max columns upto 500
df.dtypes

Country                             object
Year                                 int64
Status                              object
Life expectancy                    float64
Adult Mortality                    float64
infant deaths                        int64
Alcohol                            float64
percentage expenditure             float64
Hepatitis B                        float64
Measles                              int64
 BMI                               float64
under-five deaths                    int64
Polio                              float64
Total expenditure                  float64
Diphtheria                         float64
 HIV/AIDS                          float64
GDP                                float64
Population                         float64
 thinness  1-19 years              float64
 thinness 5-9 years                float64
Income composition of resources    float64
Schooling                          float64
dtype: object

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2938 entries, 0 to 2937
Data columns (total 22 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Country                          2938 non-null   object 
 1   Year                             2938 non-null   int64  
 2   Status                           2938 non-null   object 
 3   Life expectancy                  2928 non-null   float64
 4   Adult Mortality                  2928 non-null   float64
 5   infant deaths                    2938 non-null   int64  
 6   Alcohol                          2744 non-null   float64
 7   percentage expenditure           2938 non-null   float64
 8   Hepatitis B                      2385 non-null   float64
 9   Measles                          2938 non-null   int64  
 10   BMI                             2904 non-null   float64
 11  under-five deaths                2938 non-null   int64  
 12  Polio               

### A3. Extract the number of rows and columns present in the data

In [ ]:
# Get the shape of df
df.shape

print("Number of Rows : ", df.shape[0])
print("Number of Columns : ", df.shape[1])

Number of Rows :  2938
Number of Columns :  22


### A4. Does any of columns have extra trailing spaces in the name? If so, remove the extra spaces

In [ ]:
# Print the column names of df
df.columns

Index(['Country', 'Year', 'Status', 'Life expectancy ', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'Measles ', ' BMI ', 'under-five deaths ', 'Polio', 'Total expenditure',
       'Diphtheria ', ' HIV/AIDS', 'GDP', 'Population',
       ' thinness  1-19 years', ' thinness 5-9 years',
       'Income composition of resources', 'Schooling'],
      dtype='object')

__Inference__
- Few of the columns have extra spaces in the names like BMI, HIV/AIDS etc. 

Let's remove these

In [ ]:
# Remove the extra space from column names
df = df.rename(columns=lambda x: x.strip())

df.columns

Index(['Country', 'Year', 'Status', 'Life expectancy', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'Measles', 'BMI', 'under-five deaths', 'Polio', 'Total expenditure',
       'Diphtheria', 'HIV/AIDS', 'GDP', 'Population', 'thinness  1-19 years',
       'thinness 5-9 years', 'Income composition of resources', 'Schooling'],
      dtype='object')

### A5. Get the summary of the data

In [ ]:
# Get the summary of df
df.describe()

Year  Life expectancy  Adult Mortality  infant deaths  \
count  2938.000000      2928.000000      2928.000000    2938.000000   
mean   2007.518720        69.224932       164.796448      30.303948   
std       4.613841         9.523867       124.292079     117.926501   
min    2000.000000        36.300000         1.000000       0.000000   
25%    2004.000000        63.100000        74.000000       0.000000   
50%    2008.000000        72.100000       144.000000       3.000000   
75%    2012.000000        75.700000       228.000000      22.000000   
max    2015.000000        89.000000       723.000000    1800.000000   

           Alcohol  percentage expenditure  Hepatitis B        Measles  \
count  2744.000000             2938.000000  2385.000000    2938.000000   
mean      4.602861              738.251295    80.940461    2419.592240   
std       4.052413             1987.914858    25.070016   11467.272489   
min       0.010000                0.000000     1.000000       0.000000   
25%       0.877500                4.685343    77.000000       0.000000   
50%       3.755000               64.912906    92.000000      17.000000   
75%       7.702500              441.534144    97.000000     360.250000   
max      17.870000            19479.911610    99.000000  212183.000000   

               BMI  under-five deaths        Polio  Total expenditure  \
count  2904.000000        2938.000000  2919.000000         2712.00000   
mean     38.321247          42.035739    82.550188            5.93819   
std      20.044034         160.445548    23.428046            2.49832   
min       1.000000           0.000000     3.000000            0.37000   
25%      19.300000           0.000000    78.000000            4.26000   
50%      43.500000           4.000000    93.000000            5.75500   
75%      56.200000          28.000000    97.000000            7.49250   
max      87.300000        2500.000000    99.000000           17.60000   

        Diphtheria     HIV/AIDS            GDP    Population  \
count  2919.000000  2938.000000    2490.000000  2.286000e+03   
mean     82.324084     1.742103    7483.158469  1.275338e+07   
std      23.716912     5.077785   14270.169342  6.101210e+07   
min       2.000000     0.100000       1.681350  3.400000e+01   
25%      78.000000     0.100000     463.935626  1.957932e+05   
50%      93.000000     0.100000    1766.947595  1.386542e+06   
75%      97.000000     0.800000    5910.806335  7.420359e+06   
max      99.000000    50.600000  119172.741800  1.293859e+09   

       thinness  1-19 years  thinness 5-9 years  \
count           2904.000000         2904.000000   
mean               4.839704            4.870317   
std                4.420195            4.508882   
min                0.100000            0.100000   
25%                1.600000            1.500000   
50%                3.300000            3.300000   
75%                7.200000            7.200000   
max               27.700000           28.600000   

       Income composition of resources    Schooling  
count                      2771.000000  2775.000000  
mean                          0.627551    11.992793  
std                           0.210904     3.358920  
min                           0.000000     0.000000  
25%                           0.493000    10.100000  
50%                           0.677000    12.300000  
75%                           0.779000    14.300000  
max                           0.948000    20.700000

### A6. Are there any missing values in the data? If so, which columns have more than 10% missing values?

In [ ]:
# Missing values in the data
df_miss = df.isna().sum().sort_values(ascending=False)
df_miss

Population                         652
Hepatitis B                        553
GDP                                448
Total expenditure                  226
Alcohol                            194
Income composition of resources    167
Schooling                          163
thinness 5-9 years                  34
thinness  1-19 years                34
BMI                                 34
Polio                               19
Diphtheria                          19
Life expectancy                     10
Adult Mortality                     10
HIV/AIDS                             0
Country                              0
Year                                 0
Measles                              0
percentage expenditure               0
infant deaths                        0
Status                               0
under-five deaths                    0
dtype: int64

In [ ]:
df_miss = df_miss/df_miss.sum()
df_miss

Population                         0.254389
Hepatitis B                        0.215763
GDP                                0.174795
Total expenditure                  0.088178
Alcohol                            0.075693
Income composition of resources    0.065158
Schooling                          0.063597
thinness 5-9 years                 0.013266
thinness  1-19 years               0.013266
BMI                                0.013266
Polio                              0.007413
Diphtheria                         0.007413
Life expectancy                    0.003902
Adult Mortality                    0.003902
HIV/AIDS                           0.000000
Country                            0.000000
Year                               0.000000
Measles                            0.000000
percentage expenditure             0.000000
infant deaths                      0.000000
Status                             0.000000
under-five deaths                  0.000000
dtype: float64

__Inference__
- There are multiple columns having missing values  

Let's get the column which has more than 10% missing

In [ ]:
df_miss[df_miss > .1]

Population     0.254389
Hepatitis B    0.215763
GDP            0.174795
dtype: float64

### A7. Treat the missing values by replacing them with the column median

Replace missing values with the median.

In [ ]:
# Replace missing values using median. Except country and status column as they are categorical columns and doesn't have any missing values
for i in df.columns.drop(['Country', 'Status']):
    df[i].fillna(df[i].median(), inplace = True)

In [ ]:
df.isna().sum()

Country                            0
Year                               0
Status                             0
Life expectancy                    0
Adult Mortality                    0
infant deaths                      0
Alcohol                            0
percentage expenditure             0
Hepatitis B                        0
Measles                            0
BMI                                0
under-five deaths                  0
Polio                              0
Total expenditure                  0
Diphtheria                         0
HIV/AIDS                           0
GDP                                0
Population                         0
thinness  1-19 years               0
thinness 5-9 years                 0
Income composition of resources    0
Schooling                          0
dtype: int64

### A8. How many countries have the status as developed and developing respectively

In [ ]:
# Get the shape of df
status_df = df[['Country','Status']].drop_duplicates()
status_df.head()

Country      Status
0           Afghanistan  Developing
16              Albania  Developing
32              Algeria  Developing
48               Angola  Developing
64  Antigua and Barbuda  Developing

In [ ]:
status_df['Status'].value_counts()

Developing    161
Developed      32
Name: Status, dtype: int64

__Inference__
- 161 countries are developing and 32 developed 

### A9. Which country has the highest BMI (average)

In [ ]:
# Get the average BMI of all countires
bmi_df = df.groupby('Country').agg({'BMI': 'mean'}).reset_index()

bmi_df.head()

Country       BMI
0          Afghanistan  15.51875
1              Albania  49.06875
2              Algeria  48.74375
3               Angola  18.01875
4  Antigua and Barbuda  38.42500

In [ ]:
# Arrange the dataframe by value of BMI
bmi_df = bmi_df.sort_values(by=['BMI'], ascending=False).reset_index(drop = True)

In [ ]:
bmi_df

Country      BMI
0                               Nauru  87.3000
1                               Palau  83.3000
2                        Cook Islands  82.8000
3                    Marshall Islands  81.6000
4                              Tuvalu  79.3000
..                                ...      ...
188                       Timor-Leste  14.5500
189  Lao People's Democratic Republic  14.3625
190                        Bangladesh  12.8750
191                          Viet Nam  11.1875
192             Saint Kitts and Nevis   5.2000

[193 rows x 2 columns]

In [ ]:
bmi_df['Country'][0]

'Nauru'

__Inference__
- Nauru has the highest BMI

### A10. Who has higher life expectancy, developed or developing countries?

In [ ]:
# Get the average life expectancy among developed and developing countries
bmi_df = df.groupby('Status').agg({'Life expectancy': 'mean'}).reset_index()

bmi_df.head()

Status  Life expectancy
0   Developed        79.197852
1  Developing        67.132028

__Inference__
- Developed countries have high life expectancy (79.2) as compared to developing countires (67.1)

### A11. Which year experienced the lowest adult mortality

In [ ]:
# Get the average life expectancy among developed and developing countries
adult_mortality_by_year = df.groupby('Year').agg({'Adult Mortality': 'mean'}).reset_index()
adult_mortality_by_year

Year  Adult Mortality
0   2000       181.475410
1   2001       175.169399
2   2002       171.437158
3   2003       169.131148
4   2004       186.262295
5   2005       167.633880
6   2006       170.404372
7   2007       159.808743
8   2008       174.519126
9   2009       163.393443
10  2010       161.863388
11  2011       156.535519
12  2012       148.792350
13  2013       148.518135
14  2014       148.688525
15  2015       152.863388

In [ ]:
adult_mortality_by_year.loc[adult_mortality_by_year['Adult Mortality'] == adult_mortality_by_year['Adult Mortality'].min(), 'Year']

13    2013
Name: Year, dtype: int64

### A12. What is the average life expectancy when the GDP is above 600

In [ ]:
# Get the rows where GDP is above 600
high_gdp_df = df[df['GDP'] > 600]

high_gdp_df['Life expectancy'].mean()

71.58254910918227

### Resources


1. Practice Pandas: 
   - 101 Pandas Exercises: https://www.machinelearningplus.com/python/101-pandas-exercises-python/
   - More Exercises: https://github.com/guipsamora/pandas_exercises
2. Modern Pandas: http://tomaugspurger.github.io/modern-1-intro.html
3. Contribute to Pandas: https://github.com/pandas-dev/pandas
